1.22/st-experimental-connection/pages/04_🗂️_S3,_GCS,_and_cloud_files.py

```diff
@@ -0,0 +1,333 @@
+            
+            try:
+                _ = conn.read(parquet_file, input_format='parquet')
+            except FileNotFoundError:
+                with conn.open(parquet_file, "wb") as f:
+                    df.to_parquet(f)
+
+
+    st.write("#### Text files")
+
+    with st.echo():
+        st.write(conn.read("test-files/test.txt", input_format='text'))
+
+    st.write("#### CSV Files")
+    with st.echo():
+        st.write(conn.read("test-files/test.csv", input_format='csv'))
+
+    st.write("#### Parquet Files")
+    with st.echo():
+        st.write(conn.read("test-files/test.parquet", input_format='parquet'))
+
+
+with s3:
+    st.write("### Working with S3 files")
+    st.write("Credentials are stored in secrets.toml")
+
+    st.code(
+        """
+# In secrets.toml
+[connections.s3]
+protocol = "s3"
+key = "..."
+secret = "..."
+    """,
+        language="toml",
+    )
+
+    with st.echo():
+        conn = st.experimental_connection('s3', type=FilesConnection)
+
+    with st.expander("Setup code"):
+        with st.echo():
+            text_file = "st-connection-test/test.txt"
+            csv_file = "st-connection-test/test.csv"
+            parquet_file = "st-connection-test/test.parquet"
+            try:
+                _ = conn.read(text_file, input_format='text')
+            except FileNotFoundError:
+                with conn.open(text_file, "wt") as f:
+                    f.write("This is a test")
+            
+            try:
+                _ = conn.read(csv_file, input_format='csv')
+            except FileNotFoundError:
+                with conn.open(csv_file, "wt") as f:
+                    df.to_csv(f, index=False)
+            
+            try:
+                _ = conn.read(parquet_file, input_format='parquet')
+            except FileNotFoundError:
+                with conn.open(parquet_file, "wb") as f:
+                    df.to_parquet(f)
+
+    st.write("#### Text files")
+
+    with st.echo():
+        st.write(conn.read("s3://st-connection-test/test.txt", input_format='text'))
+
+    st.write("#### CSV Files")
+    with st.echo():
+        st.write(conn.read("s3://st-connection-test/test.csv", input_format='csv'))
+
+    st.write("#### Parquet Files")
+    with st.echo():
+        st.write(conn.read("s3://st-connection-test/test.parquet", input_format='parquet'))
+    
+    st.write("#### List operations")
+    with st.echo():
+        st.write(conn.fs.ls("s3://st-connection-test/"))
+
+with s3_other:
+    st.write("### Using S3 with your own credentials")
+
+    # HACK to get the environment variables set
+    secrets = st.secrets["connections"]["s3"]
+
+    os.environ["AWS_ACCESS_KEY_ID"] = secrets["key"]
+    os.environ["AWS_SECRET_ACCESS_KEY"] = secrets["secret"]
+
+    st.write(
+        "You can also use credentials stored in `~/.aws/config` or `AWS_ACCESS_KEY_ID` & "
+        "`AWS_SECRET_ACCES_KEY` environment variables. This example uses that approach"
```

Comment:
There's a missing space at the end of this line (that's not obvious because of the Python implicit string concatenation being used). I can make a change to fix this